File: integration/ts/auction.test.ts
/**
 * Auction integration test — stateful contract (SDK Deploy path).
 *
 * Auction is a StatefulSmartContract with properties:
 *   - auctioneer: PubKey (readonly)
 *   - highestBidder: PubKey (mutable)
 *   - highestBid: bigint (mutable)
 *   - deadline: bigint (readonly)
 *
 * Methods:
 *   - bid(sig: Sig, bidder: PubKey, bidAmount: bigint) — requires bidder's Sig + extractLocktime
 *   - close(sig: Sig) — requires auctioneer's Sig
 *
 * The bid() method requires the bidder's signature (prevents griefing) and checks
 * extractLocktime(this.txPreimage) < this.deadline, which constrains the
 * transaction's nLockTime. The close() method requires the auctioneer's Sig. Both
 * bid paths are complex enough to warrant raw tx construction for spending. We test
 * compile + deploy via the SDK. Full spending tests are covered by the Go
 * integration suite (auction_test.go).
 */

import { describe, it, expect } from 'vitest';
import { compileContract } from './helpers/compile.js';
import { RunarContract } from 'runar-sdk';
import { createFundedWallet, createWallet } from './helpers/wallet.js';
import { createProvider } from './helpers/node.js';

describe('Auction', () => {
  it('should compile the Auction contract', () => {
    const artifact = compileContract('examples/ts/auction/Auction.runar.ts');
    expect(artifact).toBeTruthy();
    expect(artifact.contractName).toBe('Auction');
  });

  it('should deploy with auctioneer, initial bidder, bid, and deadline', async () => {
    const artifact = compileContract('examples/ts/auction/Auction.runar.ts');

    const provider = createProvider();
    const auctioneer = createWallet();
    const initialBidder = createWallet();
    const { signer } = await createFundedWallet(provider);

    // Constructor: (auctioneer: PubKey, highestBidder: PubKey, highestBid: bigint, deadline: bigint)
    const contract = new RunarContract(artifact, [
      auctioneer.pubKeyHex,
      initialBidder.pubKeyHex,
      1000n,
      1000000n, // deadline far in the future
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
    expect(typeof deployTxid).toBe('string');
    expect(deployTxid.length).toBe(64);
  });

  it('should deploy with zero initial bid', async () => {
    const artifact = compileContract('examples/ts/auction/Auction.runar.ts');

    const provider = createProvider();
    const auctioneer = createWallet();
    const initialBidder = createWallet();
    const { signer } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      auctioneer.pubKeyHex,
      initialBidder.pubKeyHex,
      0n,
      500000n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
  });

  it('should deploy with same key as auctioneer and initial bidder', async () => {
    const artifact = compileContract('examples/ts/auction/Auction.runar.ts');

    const provider = createProvider();
    const auctioneerAndBidder = createWallet();
    const { signer } = await createFundedWallet(provider);

    // Same key for both roles
    const contract = new RunarContract(artifact, [
      auctioneerAndBidder.pubKeyHex,
      auctioneerAndBidder.pubKeyHex,
      500n,
      999999n,
    ]);

    const { txid: deployTxid } = await contract.deploy(provider, signer, {});
    expect(deployTxid).toBeTruthy();
  });

  // NOTE: bid() spending tests require raw transaction construction because
  // extractLocktime checks nLockTime against the deadline. The Go integration
  // tests cover the bid scenario.

  it('should close the auction with auctioneer signature', async () => {
    const artifact = compileContract('examples/ts/auction/Auction.runar.ts');

    const provider = createProvider();
    const initialBidder = createWallet();
    const { signer, pubKeyHex } = await createFundedWallet(provider);

    // Auctioneer is the funded signer so null Sig auto-computes correctly.
    // deadline=0 so extractLocktime(txPreimage) >= deadline passes with nLocktime=0
    const contract = new RunarContract(artifact, [
      pubKeyHex,
      initialBidder.pubKeyHex,
      1000n,
      0n,
    ]);

    await contract.deploy(provider, signer, {});

    // null Sig is auto-computed from the signer (who is the auctioneer)
    // close() does not continue state, so no newState needed
    const { txid: callTxid } = await contract.call(
      'close', [null], provider, signer,
    );
    expect(callTxid).toBeTruthy();
    expect(callTxid.length).toBe(64);
  });

  it('should reject close with wrong signer', async () => {
    const artifact = compileContract('examples/ts/auction/Auction.runar.ts');

    const provider = createProvider();
    const initialBidder = createWallet();
    // Deploy with auctioneer=walletA
    const { signer: auctioneerSigner, pubKeyHex: auctioneerPubKey } = await createFundedWallet(provider);

    const contract = new RunarContract(artifact, [
      auctioneerPubKey,
      initialBidder.pubKeyHex,
      1000n,
      0n, // deadline=0 so extractLocktime passes
    ]);

    await contract.deploy(provider, auctioneerSigner, {});

    // Call close with walletB — checkSig will fail on-chain
    const { signer: wrongSigner } = await createFundedWallet(provider);

    await expect(
      contract.call('close', [null], provider, wrongSigner),
    ).rejects.toThrow();
  });
});

File: integration/zig/src/auction_test.zig
const std = @import("std");
const runar = @import("runar");
const helpers = @import("helpers.zig");
const compile = @import("compile.zig");

test "Auction_Compile" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/auction/Auction.runar.zig") catch |err| {
        std.log.warn("Could not compile Auction contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    try std.testing.expectEqualStrings("Auction", artifact.contract_name);
    try std.testing.expect(artifact.isStateful());
    std.log.info("Auction compiled: {d} bytes", .{artifact.script.len / 2});
}

test "Auction_Deploy" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/auction/Auction.runar.zig") catch |err| {
        std.log.warn("Could not compile Auction contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var auctioneer = try helpers.newWallet(allocator);
    defer auctioneer.deinit();
    var bidder = try helpers.newWallet(allocator);
    defer bidder.deinit();

    const auctioneer_pk = try auctioneer.pubKeyHex(allocator);
    defer allocator.free(auctioneer_pk);
    const bidder_pk = try bidder.pubKeyHex(allocator);
    defer allocator.free(bidder_pk);

    // Constructor params: auctioneer, highestBidder, highestBid, deadline
    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = auctioneer_pk },
        .{ .bytes = bidder_pk },
        .{ .int = 1000 },
        .{ .int = 1000000 },
    });
    defer contract.deinit();

    var funder = try helpers.newWallet(allocator);
    defer funder.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &funder, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try funder.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(deploy_txid);

    try std.testing.expectEqual(@as(usize, 64), deploy_txid.len);
    std.log.info("Auction deployed: {s}", .{deploy_txid});
}

test "Auction_DeployZeroBid" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/auction/Auction.runar.zig") catch |err| {
        std.log.warn("Could not compile Auction contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var auctioneer = try helpers.newWallet(allocator);
    defer auctioneer.deinit();
    var bidder = try helpers.newWallet(allocator);
    defer bidder.deinit();

    const auctioneer_pk = try auctioneer.pubKeyHex(allocator);
    defer allocator.free(auctioneer_pk);
    const bidder_pk = try bidder.pubKeyHex(allocator);
    defer allocator.free(bidder_pk);

    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = auctioneer_pk },
        .{ .bytes = bidder_pk },
        .{ .int = 0 },
        .{ .int = 1000000 },
    });
    defer contract.deinit();

    var funder = try helpers.newWallet(allocator);
    defer funder.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &funder, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try funder.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(deploy_txid);

    const utxo = contract.getCurrentUtxo();
    try std.testing.expect(utxo != null);
    std.log.info("Auction deployed with zero bid: {s}", .{deploy_txid});
}

test "Auction_DeploySameKey" {
    const allocator = std.testing.allocator;

    if (!helpers.isNodeAvailable(allocator)) {
        std.log.warn("Regtest node not available, skipping test", .{});
        return;
    }

    var artifact = compile.compileContract(allocator, "examples/zig/auction/Auction.runar.zig") catch |err| {
        std.log.warn("Could not compile Auction contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    var wallet = try helpers.newWallet(allocator);
    defer wallet.deinit();

    const pk_hex = try wallet.pubKeyHex(allocator);
    defer allocator.free(pk_hex);

    // Same key as auctioneer and bidder
    var contract = try runar.RunarContract.init(allocator, &artifact, &[_]runar.StateValue{
        .{ .bytes = pk_hex },
        .{ .bytes = pk_hex },
        .{ .int = 1000 },
        .{ .int = 1000000 },
    });
    defer contract.deinit();

    var funder = try helpers.newWallet(allocator);
    defer funder.deinit();
    const fund_txid = try helpers.fundWallet(allocator, &funder, 1.0);
    defer allocator.free(fund_txid);

    var rpc_provider = helpers.RPCProvider.init(allocator);
    var local_signer = try funder.localSigner();

    const deploy_txid = try contract.deploy(rpc_provider.provider(), local_signer.signer(), .{ .satoshis = 5000 });
    defer allocator.free(deploy_txid);

    const utxo = contract.getCurrentUtxo();
    try std.testing.expect(utxo != null);
    std.log.info("Auction deployed with same key for auctioneer and bidder: {s}", .{deploy_txid});
}

test "Auction_ABI_Methods" {
    const allocator = std.testing.allocator;

    var artifact = compile.compileContract(allocator, "examples/zig/auction/Auction.runar.zig") catch |err| {
        std.log.warn("Could not compile Auction contract: {any}, skipping test", .{err});
        return;
    };
    defer artifact.deinit();

    // Auction should have public methods including bid and close
    var public_count: usize = 0;
    for (artifact.abi.methods) |m| {
        if (m.is_public) public_count += 1;
    }
    try std.testing.expect(public_count >= 2);
}
